{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "54f821bb-0d1d-4acb-ba2a-c5fa909dd608",
   "metadata": {},
   "source": [
    "# Describe study metadata\n",
    "\n",
    "Each screen contains an experiment with different parameters and conditions.\n",
    "\n",
    "Extract this information based on ID and save details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef413c2b-421c-4c0c-adfb-9378af1caa50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import time\n",
    "import os\n",
    "import requests\n",
    "import pandas as pd\n",
    "import pyarrow as pa\n",
    "import pyarrow.parquet as pq\n",
    "import multiprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "58e4eaf0-71e1-48cf-bd9e-1ef743c9f09f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_study_info(session, screen_id):\n",
    "    \"\"\"Pull metadata info per screen, given screen id\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    session: Requests.session()\n",
    "        Requests session providing access to IDR API\n",
    "    screen_id: int\n",
    "        ID of the screen data set\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame() of metadata per screen id\n",
    "    \"\"\"\n",
    "    base_url = \"https://idr.openmicroscopy.org/webclient/api/annotations/\"\n",
    "    screen_url = f\"?type=map&screen={screen_id}\"\n",
    "\n",
    "    url = f\"{base_url}{screen_url}\"\n",
    "    response = session.get(url).json()\n",
    "\n",
    "    annotations = response[\"annotations\"]\n",
    "    study_index = [x[\"ns\"] for x in annotations].index(\n",
    "        'idr.openmicroscopy.org/study/info')\n",
    "\n",
    "    id_ = annotations[study_index][\"id\"]\n",
    "    date_ = annotations[study_index][\"date\"]\n",
    "    name_ = annotations[study_index][\"link\"][\"parent\"][\"name\"]\n",
    "\n",
    "    details_ = (\n",
    "        pd.DataFrame({x[0]: x[1]\n",
    "                      for x in annotations[study_index][\"values\"]}, index=[0])\n",
    "        .assign(\n",
    "            internal_id=id_,\n",
    "            upload_date=date_,\n",
    "            idr_name=name_,\n",
    "            screen_id=screen_id\n",
    "        )\n",
    "    )\n",
    "\n",
    "    return details_\n",
    "\n",
    "\n",
    "def describe_screen(screen_id):\n",
    "    \"\"\"Pull additional metadata info per plate, given screen id\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    screen_id: int\n",
    "        ID of the screen data set\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame() of the following metadata values:\n",
    "\n",
    "        screen_id: IDR ID for the screen\n",
    "        plate_id: IDR ID for each plate\n",
    "        plate_name: Names given to each plate\n",
    "        image_id: IDR ID for each image\n",
    "        cell_line: Cell line used in the screen experiment\n",
    "        strain: Strain of the cell line (if specified)\n",
    "        gene_identifier: Accession code for the gene being perturbed in a well\n",
    "        phenotype_identifier: Accession code for the phenotype perturbed in a well\n",
    "        stain: Set of the stains used in the screen\n",
    "        stain_target: The target protein or media of the stain\n",
    "        pixel_size_x: Width of the image\n",
    "        pixel_size_y: Height of the image\n",
    "        imaging_method: Method used to collect images (ex: fluorescence microscopy)\n",
    "    \"\"\"\n",
    "    session = requests.Session()\n",
    "\n",
    "    # Get number of plates per screen and append to a dictionary\n",
    "    PLATES_URL = f\"https://idr.openmicroscopy.org/webclient/api/plates/?id={screen_id}\"\n",
    "    all_plates = session.get(PLATES_URL).json()[\"plates\"]\n",
    "    study_plates = {x[\"id\"]: x[\"name\"] for x in all_plates}\n",
    "    print(f\"Number of plates found in screen {screen_id}: \", len(study_plates))\n",
    "\n",
    "    # Initialize results list\n",
    "    plate_results = list()\n",
    "\n",
    "    # Iterate through all plates in the study\n",
    "    for plate in study_plates:\n",
    "        imageIDs = list()\n",
    "        plate_name = study_plates[plate]\n",
    "\n",
    "        # Access .json file for the plate ID. Contains image ID (id) numbers\n",
    "        # for replicate images per plate.\n",
    "        WELLS_IMAGES_URL = f\"https://idr.openmicroscopy.org/webgateway/plate/{plate}/\"\n",
    "        grid = session.get(WELLS_IMAGES_URL).json()\n",
    "\n",
    "        try:\n",
    "            pixel_size_x = grid[\"image_sizes\"][0][\"x\"]\n",
    "            pixel_size_y = grid[\"image_sizes\"][0][\"y\"]\n",
    "\n",
    "            # Get image and well ids\n",
    "            for image in grid['grid'][0]:\n",
    "                # Append IDs to iterable lists\n",
    "                thumb_url = image['thumb_url'].rstrip(image['thumb_url'][-1])\n",
    "                image_id = thumb_url.split('/')[-1]\n",
    "                imageIDs.append(image_id)\n",
    "\n",
    "        except (ValueError, KeyError):\n",
    "            plate_results.append(\n",
    "                [screen_id, plate, plate_name, None, None, None, None, None])\n",
    "            continue\n",
    "\n",
    "        # Get image details from each image id for each plate\n",
    "        for id in imageIDs:\n",
    "            MAP_URL = f\"https://idr.openmicroscopy.org/webclient/api/annotations/?type=map&image={id}\"\n",
    "            annotations = session.get(MAP_URL).json()[\"annotations\"]\n",
    "\n",
    "            # Get stain and stain target\n",
    "            try:\n",
    "                bulk_index = [x[\"ns\"] for x in annotations].index(\n",
    "                    'openmicroscopy.org/omero/bulk_annotations')\n",
    "                channels = {x[0]: x[1]\n",
    "                            for x in annotations[bulk_index]\n",
    "                            [\"values\"]}[\"Channels\"]\n",
    "\n",
    "                # Clean channels value and separate into stain and stain target\n",
    "                stain = set()\n",
    "                stain_target = set()\n",
    "                for entry in channels.split(\";\"):\n",
    "                    temp_list = entry.split(\":\")\n",
    "                    stain.add(temp_list[0])\n",
    "                    stain_target.add(temp_list[1])\n",
    "\n",
    "            except (ValueError, KeyError):\n",
    "                stain = \"Not listed\"\n",
    "                stain_target = \"Not listed\"\n",
    "\n",
    "            # Get cell line\n",
    "            try:\n",
    "                cell_line_index = [x[\"ns\"] for x in annotations].index(\n",
    "                    'openmicroscopy.org/mapr/cell_line')\n",
    "                cell_line = {x[0]: x[1] for x in\n",
    "                             annotations[cell_line_index][\"values\"]}[\"Cell Line\"]\n",
    "            except (ValueError, KeyError):\n",
    "                cell_line = \"Not listed\"\n",
    "\n",
    "            # Get strain of cell line\n",
    "            try:\n",
    "                strain = {x[0]: x[1] for x in\n",
    "                          annotations[bulk_index][\"values\"]}[\"Strain\"]\n",
    "            except (ValueError, KeyError):\n",
    "                strain = \"Not listed\"\n",
    "\n",
    "            # Get gene identification accession code\n",
    "            try:\n",
    "                gene_identifier_index = int(\n",
    "                    [x[\"ns\"] for x in annotations].index(\n",
    "                        'openmicroscopy.org/mapr/gene'))\n",
    "                gene_identifier = {x[0]: x[1] for x in annotations[gene_identifier_index][\"values\"]}[\n",
    "                    \"Gene Identifier\"]\n",
    "            except (ValueError, KeyError):\n",
    "                gene_identifier = \"Not Listed\"\n",
    "\n",
    "            # Get phenotype identification accession code\n",
    "            try:\n",
    "                phenotype_identifier_index = int(\n",
    "                    [x[\"ns\"] for x in annotations].index('openmicroscopy.org/mapr/phenotype'))\n",
    "                phenotype_identifier = {x[0]: x[1] for x in annotations[phenotype_identifier_index][\"values\"]}[\n",
    "                    \"Phenotype Term Accession\"]\n",
    "            except (ValueError, KeyError):\n",
    "                phenotype_identifier = \"Not Listed\"\n",
    "\n",
    "            # Build results\n",
    "            plate_results.append([\n",
    "                screen_id,\n",
    "                plate,\n",
    "                plate_name,\n",
    "                id,\n",
    "                cell_line,\n",
    "                strain,\n",
    "                gene_identifier,\n",
    "                phenotype_identifier,\n",
    "                stain,\n",
    "                stain_target,\n",
    "                pixel_size_x,\n",
    "                pixel_size_y\n",
    "            ])\n",
    "\n",
    "    # Output results\n",
    "    plate_results_df = pd.DataFrame(\n",
    "        plate_results,\n",
    "        columns=[\n",
    "            \"screen_id\",\n",
    "            \"plate_id\",\n",
    "            \"plate_name\",\n",
    "            \"image_id\",\n",
    "            \"cell_line\",\n",
    "            \"strain\",\n",
    "            \"gene_identifier\",\n",
    "            \"phenotype_identifier\",\n",
    "            \"stain\",\n",
    "            \"stain_target\",\n",
    "            \"pixel_size_x\",\n",
    "            \"pixel_size_y\"\n",
    "        ]\n",
    "    )\n",
    "    return plate_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abe51937-9e13-41e4-85a4-9834174009f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = pathlib.Path(\"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7db87421-4204-4732-bd14-c52e068bc129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(198, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>title</th>\n",
       "      <th>description</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>idr0001-graml-sysgro/screenA</td>\n",
       "      <td>A genomic Multiprocess survey of machineries t...</td>\n",
       "      <td>Primary screen of fission yeast knock out muta...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102</td>\n",
       "      <td>idr0002-heriche-condensation/screenA</td>\n",
       "      <td>Integration of biological data by kernels on g...</td>\n",
       "      <td>Screen of 100 candidate genes predicted to be ...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51</td>\n",
       "      <td>idr0003-breker-plasticity/screenA</td>\n",
       "      <td>A novel single-cell screening platform reveals...</td>\n",
       "      <td>Screen to characterize yeast stress responses ...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>202</td>\n",
       "      <td>idr0004-thorpe-rad52/screenA</td>\n",
       "      <td>Bringing Rad52 foci into focus.</td>\n",
       "      <td>This screen assesses the proportion of cells c...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>597</td>\n",
       "      <td>idr0005-toret-adhesion/screenA</td>\n",
       "      <td>A genome-wide screen identifies conserved prot...</td>\n",
       "      <td>A genome-wide RNAi screen for loss of DE- cadh...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>751</td>\n",
       "      <td>idr0005-toret-adhesion/screenB</td>\n",
       "      <td>A genome-wide screen identifies conserved prot...</td>\n",
       "      <td>Re-screen of 803 hits from the first screen, u...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>253</td>\n",
       "      <td>idr0006-fong-nuclearbodies/screenA</td>\n",
       "      <td>Whole-genome screening identifies proteins loc...</td>\n",
       "      <td>A whole-genome screening for proteins localize...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>201</td>\n",
       "      <td>idr0007-srikumar-sumo/screenA</td>\n",
       "      <td>Global analysis of SUMO chain function reveals...</td>\n",
       "      <td>In this screen two mutant yeast strains unable...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>154</td>\n",
       "      <td>idr0008-rohn-actinome/screenA</td>\n",
       "      <td>Comparative RNAi screening identifies a conser...</td>\n",
       "      <td>A genome-wide, high-content RNAi screen in the...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>206</td>\n",
       "      <td>idr0008-rohn-actinome/screenB</td>\n",
       "      <td>Comparative RNAi screening identifies a conser...</td>\n",
       "      <td>This screen consists of siRNAs targetting 516 ...</td>\n",
       "      <td>Screen</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id                                  name  \\\n",
       "0    3          idr0001-graml-sysgro/screenA   \n",
       "1  102  idr0002-heriche-condensation/screenA   \n",
       "2   51     idr0003-breker-plasticity/screenA   \n",
       "3  202          idr0004-thorpe-rad52/screenA   \n",
       "4  597        idr0005-toret-adhesion/screenA   \n",
       "5  751        idr0005-toret-adhesion/screenB   \n",
       "6  253    idr0006-fong-nuclearbodies/screenA   \n",
       "7  201         idr0007-srikumar-sumo/screenA   \n",
       "8  154         idr0008-rohn-actinome/screenA   \n",
       "9  206         idr0008-rohn-actinome/screenB   \n",
       "\n",
       "                                               title  \\\n",
       "0  A genomic Multiprocess survey of machineries t...   \n",
       "1  Integration of biological data by kernels on g...   \n",
       "2  A novel single-cell screening platform reveals...   \n",
       "3                    Bringing Rad52 foci into focus.   \n",
       "4  A genome-wide screen identifies conserved prot...   \n",
       "5  A genome-wide screen identifies conserved prot...   \n",
       "6  Whole-genome screening identifies proteins loc...   \n",
       "7  Global analysis of SUMO chain function reveals...   \n",
       "8  Comparative RNAi screening identifies a conser...   \n",
       "9  Comparative RNAi screening identifies a conser...   \n",
       "\n",
       "                                         description category  \n",
       "0  Primary screen of fission yeast knock out muta...   Screen  \n",
       "1  Screen of 100 candidate genes predicted to be ...   Screen  \n",
       "2  Screen to characterize yeast stress responses ...   Screen  \n",
       "3  This screen assesses the proportion of cells c...   Screen  \n",
       "4  A genome-wide RNAi screen for loss of DE- cadh...   Screen  \n",
       "5  Re-screen of 803 hits from the first screen, u...   Screen  \n",
       "6  A whole-genome screening for proteins localize...   Screen  \n",
       "7  In this screen two mutant yeast strains unable...   Screen  \n",
       "8  A genome-wide, high-content RNAi screen in the...   Screen  \n",
       "9  This screen consists of siRNAs targetting 516 ...   Screen  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load IDR ids\n",
    "data_dir = pathlib.Path(\n",
    "    \"~/Documents/publicly-available-microscopy-data/IDR/data\")\n",
    "id_file = pathlib.Path(data_dir, \"idr_ids.tsv\")\n",
    "id_df = pd.read_csv(id_file, sep=\"\\t\")\n",
    "\n",
    "print(id_df.shape)\n",
    "id_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "32061d5e-6d36-4432-98ab-93383c3e08b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create http session\n",
    "INDEX_PAGE = \"https://idr.openmicroscopy.org/webclient/?experimenter=-1\"\n",
    "with requests.Session() as session:\n",
    "    request = requests.Request('GET', INDEX_PAGE)\n",
    "    prepped = session.prepare_request(request)\n",
    "    response = session.send(prepped)\n",
    "    if response.status_code != 200:\n",
    "        response.raise_for_status()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f4926c7b-e4ff-4c2a-b9b4-49163adb70fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are a total of 80 screens\n",
      "(80, 24)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Study Type</th>\n",
       "      <th>Screen Type</th>\n",
       "      <th>Screen Technology Type</th>\n",
       "      <th>Imaging Method</th>\n",
       "      <th>Publication Title</th>\n",
       "      <th>Publication Authors</th>\n",
       "      <th>PubMed ID</th>\n",
       "      <th>PMC ID</th>\n",
       "      <th>...</th>\n",
       "      <th>External URL</th>\n",
       "      <th>Annotation File</th>\n",
       "      <th>internal_id</th>\n",
       "      <th>upload_date</th>\n",
       "      <th>idr_name</th>\n",
       "      <th>screen_id</th>\n",
       "      <th>Study Title</th>\n",
       "      <th>Data Publisher</th>\n",
       "      <th>Data DOI</th>\n",
       "      <th>BioStudies Accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cell</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>high content screen</td>\n",
       "      <td>primary screen</td>\n",
       "      <td>gene deletion screen</td>\n",
       "      <td>spinning disk confocal microscopy</td>\n",
       "      <td>A genomic Multiprocess survey of machineries t...</td>\n",
       "      <td>Graml V, Studera X, Lawson JLD, Chessel A, Gey...</td>\n",
       "      <td>25373780 https://www.ncbi.nlm.nih.gov/pubmed/2...</td>\n",
       "      <td>PMC4648281 https://www.ncbi.nlm.nih.gov/pmc/ar...</td>\n",
       "      <td>...</td>\n",
       "      <td>www.sysgro.org</td>\n",
       "      <td>idr0001-screenA-annotation.csv https://github....</td>\n",
       "      <td>20516150</td>\n",
       "      <td>2019-06-04T10:43:11+01:00</td>\n",
       "      <td>idr0001-graml-sysgro/screenA</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cell</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>high content screen</td>\n",
       "      <td>primary screen</td>\n",
       "      <td>RNAi screen</td>\n",
       "      <td>fluorescence microscopy</td>\n",
       "      <td>Integration of biological data by kernels on g...</td>\n",
       "      <td>Hériché JK, Lees JG, Morilla I, Walter T, Petr...</td>\n",
       "      <td>24943848 https://www.ncbi.nlm.nih.gov/pubmed/2...</td>\n",
       "      <td>PMC4142622 https://www.ncbi.nlm.nih.gov/pmc/ar...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>idr0002-screenA-annotation.csv https://github....</td>\n",
       "      <td>20516151</td>\n",
       "      <td>2019-06-04T10:43:19+01:00</td>\n",
       "      <td>idr0002-heriche-condensation/screenA</td>\n",
       "      <td>102</td>\n",
       "      <td>Focused mitotic chromsome condensaton screen u...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cell</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>high content screen</td>\n",
       "      <td>primary screen</td>\n",
       "      <td>protein screen</td>\n",
       "      <td>fluorescence microscopy</td>\n",
       "      <td>A novel single-cell screening platform reveals...</td>\n",
       "      <td>Breker M, Gymrek M, Schuldiner M</td>\n",
       "      <td>23509072 https://www.ncbi.nlm.nih.gov/pubmed/2...</td>\n",
       "      <td>PMC3601363 https://www.ncbi.nlm.nih.gov/pmc/ar...</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.weizmann.ac.il/molgen/loqate/</td>\n",
       "      <td>idr0003-screenA-annotation.csv https://github....</td>\n",
       "      <td>20516152</td>\n",
       "      <td>2019-06-04T10:43:22+01:00</td>\n",
       "      <td>idr0003-breker-plasticity/screenA</td>\n",
       "      <td>51</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Type                   Organism           Study Type     Screen Type  \\\n",
       "0        cell  Schizosaccharomyces pombe  high content screen  primary screen   \n",
       "1        cell               Homo sapiens  high content screen  primary screen   \n",
       "2        cell   Saccharomyces cerevisiae  high content screen  primary screen   \n",
       "\n",
       "  Screen Technology Type                     Imaging Method  \\\n",
       "0   gene deletion screen  spinning disk confocal microscopy   \n",
       "1            RNAi screen            fluorescence microscopy   \n",
       "2         protein screen            fluorescence microscopy   \n",
       "\n",
       "                                   Publication Title  \\\n",
       "0  A genomic Multiprocess survey of machineries t...   \n",
       "1  Integration of biological data by kernels on g...   \n",
       "2  A novel single-cell screening platform reveals...   \n",
       "\n",
       "                                 Publication Authors  \\\n",
       "0  Graml V, Studera X, Lawson JLD, Chessel A, Gey...   \n",
       "1  Hériché JK, Lees JG, Morilla I, Walter T, Petr...   \n",
       "2                   Breker M, Gymrek M, Schuldiner M   \n",
       "\n",
       "                                           PubMed ID  \\\n",
       "0  25373780 https://www.ncbi.nlm.nih.gov/pubmed/2...   \n",
       "1  24943848 https://www.ncbi.nlm.nih.gov/pubmed/2...   \n",
       "2  23509072 https://www.ncbi.nlm.nih.gov/pubmed/2...   \n",
       "\n",
       "                                              PMC ID  ...  \\\n",
       "0  PMC4648281 https://www.ncbi.nlm.nih.gov/pmc/ar...  ...   \n",
       "1  PMC4142622 https://www.ncbi.nlm.nih.gov/pmc/ar...  ...   \n",
       "2  PMC3601363 https://www.ncbi.nlm.nih.gov/pmc/ar...  ...   \n",
       "\n",
       "                               External URL  \\\n",
       "0                            www.sysgro.org   \n",
       "1                                       NaN   \n",
       "2  http://www.weizmann.ac.il/molgen/loqate/   \n",
       "\n",
       "                                     Annotation File internal_id  \\\n",
       "0  idr0001-screenA-annotation.csv https://github....    20516150   \n",
       "1  idr0002-screenA-annotation.csv https://github....    20516151   \n",
       "2  idr0003-screenA-annotation.csv https://github....    20516152   \n",
       "\n",
       "                 upload_date                              idr_name screen_id  \\\n",
       "0  2019-06-04T10:43:11+01:00          idr0001-graml-sysgro/screenA         3   \n",
       "1  2019-06-04T10:43:19+01:00  idr0002-heriche-condensation/screenA       102   \n",
       "2  2019-06-04T10:43:22+01:00     idr0003-breker-plasticity/screenA        51   \n",
       "\n",
       "                                         Study Title Data Publisher Data DOI  \\\n",
       "0                                                NaN            NaN      NaN   \n",
       "1  Focused mitotic chromsome condensaton screen u...            NaN      NaN   \n",
       "2                                                NaN            NaN      NaN   \n",
       "\n",
       "   BioStudies Accession  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "\n",
       "[3 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract summary details for all screens\n",
    "screen_ids = id_df.query(\"category=='Screen'\").id.tolist()\n",
    "screen_details_df = (\n",
    "    pd.concat([\n",
    "       extract_study_info(session=session, screen_id=x) for x in screen_ids\n",
    "    ], axis=\"rows\")\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "output_file = pathlib.Path(data_dir, \"screen_details.tsv\")\n",
    "screen_details_df.to_csv(output_file, index=False, sep=\"\\t\")\n",
    "\n",
    "print(screen_details_df.shape)\n",
    "screen_details_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2cea17ea-8330-464a-ab3a-81a20f5002b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now processing screen: 3\n",
      "192 plates found. Done\n",
      "\n",
      "Now processing screen: 102\n",
      "12 plates found. Done\n",
      "\n",
      "Now processing screen: 51\n",
      "85 plates found. Done\n",
      "\n",
      "Now processing screen: 202\n",
      "47 plates found. Done\n",
      "\n",
      "Now processing screen: 597\n",
      "141 plates found. Done\n",
      "\n",
      "Now processing screen: 751\n",
      "18 plates found. Done\n",
      "\n",
      "Now processing screen: 253\n",
      "169 plates found. Done\n",
      "\n",
      "Now processing screen: 201\n",
      "12 plates found. Done\n",
      "\n",
      "Now processing screen: 154\n",
      "58 plates found. Done\n",
      "\n",
      "Now processing screen: 206\n",
      "11 plates found. Done\n",
      "\n",
      "Now processing screen: 251\n",
      "500 plates found. Done\n",
      "\n",
      "Now processing screen: 803\n",
      "70 plates found. Done\n",
      "\n",
      "Now processing screen: 1351\n",
      "148 plates found. Done\n",
      "\n",
      "Now processing screen: 1501\n",
      "129 plates found. Done\n",
      "\n",
      "Now processing screen: 1551\n",
      "40 plates found. Done\n",
      "\n",
      "Now processing screen: 1601\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 1602\n",
      "8 plates found. Done\n",
      "\n",
      "Now processing screen: 1603\n",
      "1 plates found. Done\n",
      "\n",
      "Now processing screen: 1202\n",
      "68 plates found. Done\n",
      "\n",
      "Now processing screen: 1101\n",
      "500 plates found. Done\n",
      "\n",
      "Now processing screen: 1302\n",
      "28 plates found. Done\n",
      "\n",
      "Now processing screen: 1201\n",
      "84 plates found. Done\n",
      "\n",
      "Now processing screen: 1251\n",
      "413 plates found. Done\n",
      "\n",
      "Now processing screen: 1151\n",
      "96 plates found. Done\n",
      "\n",
      "Now processing screen: 1203\n",
      "7 plates found. Done\n",
      "\n",
      "Now processing screen: 1204\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 2151\n",
      "500 plates found. Done\n",
      "\n",
      "Now processing screen: 2152\n",
      "152 plates found. Done\n",
      "\n",
      "Now processing screen: 1851\n",
      "3 plates found. Done\n",
      "\n",
      "Now processing screen: 1651\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 1652\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 1653\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 1654\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 1801\n",
      "10 plates found. Done\n",
      "\n",
      "Now processing screen: 1751\n",
      "12 plates found. Done\n",
      "\n",
      "Now processing screen: 1901\n",
      "29 plates found. Done\n",
      "\n",
      "Now processing screen: 2001\n",
      "55 plates found. Done\n",
      "\n",
      "Now processing screen: 1952\n",
      "20 plates found. Done\n",
      "\n",
      "Now processing screen: 2051\n",
      "69 plates found. Done\n",
      "\n",
      "Now processing screen: 2301\n",
      "10 plates found. Done\n",
      "\n",
      "Now processing screen: 2302\n",
      "10 plates found. Done\n",
      "\n",
      "Now processing screen: 2303\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 2101\n",
      "31 plates found. Done\n",
      "\n",
      "Now processing screen: 2351\n",
      "9 plates found. Done\n",
      "\n",
      "Now processing screen: 2251\n",
      "124 plates found. Done\n",
      "\n",
      "Now processing screen: 2952\n",
      "10 plates found. Done\n",
      "\n",
      "Now processing screen: 2953\n",
      "84 plates found. Done\n",
      "\n",
      "Now processing screen: 2501\n",
      "117 plates found. Done\n",
      "\n",
      "Now processing screen: 2502\n",
      "57 plates found. Done\n",
      "\n",
      "Now processing screen: 2701\n",
      "18 plates found. Done\n",
      "\n",
      "Now processing screen: 2401\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 2402\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 2403\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 2404\n",
      "16 plates found. Done\n",
      "\n",
      "Now processing screen: 2405\n",
      "16 plates found. Done\n",
      "\n",
      "Now processing screen: 2651\n",
      "500 plates found. Done\n",
      "\n",
      "Now processing screen: 2851\n",
      "22 plates found. Done\n",
      "\n",
      "Now processing screen: 2451\n",
      "8 plates found. Done\n",
      "\n",
      "Now processing screen: 2751\n",
      "83 plates found. Done\n",
      "\n",
      "Now processing screen: 2602\n",
      "64 plates found. Done\n",
      "\n",
      "Now processing screen: 2603\n",
      "102 plates found. Done\n",
      "\n",
      "Now processing screen: 2801\n",
      "8 plates found. Done\n",
      "\n",
      "Now processing screen: 3001\n",
      "30 plates found. Done\n",
      "\n",
      "Now processing screen: 3202\n",
      "4 plates found. Done\n",
      "\n",
      "Now processing screen: 3203\n",
      "1 plates found. Done\n",
      "\n",
      "Now processing screen: 3204\n",
      "1 plates found. Done\n",
      "\n",
      "Now processing screen: 3205\n",
      "1 plates found. Done\n",
      "\n",
      "Now processing screen: 3055\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 3054\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 3053\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 3052\n",
      "16 plates found. Done\n",
      "\n",
      "Now processing screen: 3051\n",
      "16 plates found. Done\n",
      "\n",
      "Now processing screen: 3103\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 3104\n",
      "1 plates found. Done\n",
      "\n",
      "Now processing screen: 3102\n",
      "16 plates found. Done\n",
      "\n",
      "Now processing screen: 3101\n",
      "16 plates found. Done\n",
      "\n",
      "Now processing screen: 3106\n",
      "1 plates found. Done\n",
      "\n",
      "Now processing screen: 3107\n",
      "2 plates found. Done\n",
      "\n",
      "Now processing screen: 3105\n",
      "12 plates found. Done\n",
      "\n",
      "Now processing screen: 3151\n",
      "84 plates found. Done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Subset the data\n",
    "test_indices = [0, 1, 2, 3]\n",
    "test_screen_ids = [screen_ids[i] for i in test_indices]\n",
    "\n",
    "# Initialize Pool object for threading\n",
    "start = time.time()\n",
    "available_cores = len(os.sched_getaffinity(0))\n",
    "pool = multiprocessing.Pool(processes=available_cores)\n",
    "print(f\"\\nNow processing {len(test_screen_ids)} screens with {available_cores} cpu cores.\\n\")\n",
    "\n",
    "# Pull pertinent details about the screen (plates, wells, channels, cell line, etc.)\n",
    "plate_results_dfs = pool.map(describe_screen, test_screen_ids)\n",
    "\n",
    "# Terminate pool processes\n",
    "pool.close()\n",
    "pool.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94ac3418-38e0-4989-b9f2-55b1f18a646f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5217, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>screen_id</th>\n",
       "      <th>plate_id</th>\n",
       "      <th>plate_name</th>\n",
       "      <th>n_wells</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>channels</th>\n",
       "      <th>pixel_size_x</th>\n",
       "      <th>pixel_size_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>2551</td>\n",
       "      <td>JL_120731_S6A</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>2552</td>\n",
       "      <td>JL_120731_S6B</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>2554</td>\n",
       "      <td>JL_120801_S7A</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2553</td>\n",
       "      <td>JL_120801_S7B</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2555</td>\n",
       "      <td>JL_120802_S8A</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>2556</td>\n",
       "      <td>JL_120802_S8B</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>2557</td>\n",
       "      <td>JL_120803_S9A</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>2558</td>\n",
       "      <td>JL_120803_S9B</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>2559</td>\n",
       "      <td>JL_120804_S10A</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>2560</td>\n",
       "      <td>JL_120804_S10B</td>\n",
       "      <td>96</td>\n",
       "      <td>Not listed</td>\n",
       "      <td>GFP:endogenous alpha tubulin 2;Cascade blue:gr...</td>\n",
       "      <td>1376</td>\n",
       "      <td>1040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   screen_id  plate_id      plate_name  n_wells   cell_line  \\\n",
       "0          3      2551   JL_120731_S6A       96  Not listed   \n",
       "1          3      2552   JL_120731_S6B       96  Not listed   \n",
       "2          3      2554   JL_120801_S7A       96  Not listed   \n",
       "3          3      2553   JL_120801_S7B       96  Not listed   \n",
       "4          3      2555   JL_120802_S8A       96  Not listed   \n",
       "5          3      2556   JL_120802_S8B       96  Not listed   \n",
       "6          3      2557   JL_120803_S9A       96  Not listed   \n",
       "7          3      2558   JL_120803_S9B       96  Not listed   \n",
       "8          3      2559  JL_120804_S10A       96  Not listed   \n",
       "9          3      2560  JL_120804_S10B       96  Not listed   \n",
       "\n",
       "                                            channels  pixel_size_x  \\\n",
       "0  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "1  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "2  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "3  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "4  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "5  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "6  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "7  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "8  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "9  GFP:endogenous alpha tubulin 2;Cascade blue:gr...          1376   \n",
       "\n",
       "   pixel_size_y  \n",
       "0          1040  \n",
       "1          1040  \n",
       "2          1040  \n",
       "3          1040  \n",
       "4          1040  \n",
       "5          1040  \n",
       "6          1040  \n",
       "7          1040  \n",
       "8          1040  \n",
       "9          1040  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine to create full dataframe\n",
    "all_plate_results_df = pd.concat(plate_results_dfs, ignore_index=True)\n",
    "\n",
    "# Collect imaging method metadata for each screen\n",
    "img_screen_index = dict()\n",
    "for index in screen_details_df.itertuples(index=False):\n",
    "    screenID = index[19]\n",
    "    img_type = index[5]\n",
    "    img_screen_index[screenID] = img_type\n",
    "\n",
    "# Map imaging method per screen to the final data frame\n",
    "all_plate_results_df[\"imaging_method\"] = all_plate_results_df[\"screen_id\"].map(\n",
    "    img_screen_index)\n",
    "\n",
    "print(f'Metadata collected. Running cost is {(time.time()-start)/60:.1f} min. ', 'Now saving file.')\n",
    "\n",
    "# Save data frame as a single parquet file\n",
    "output_file = pathlib.Path(data_dir, \"plate_details_per_screen.parquet\")\n",
    "pq_table = pa.Table.from_pandas(all_plate_results_df)\n",
    "pq.write_table(pq_table, output_file)\n",
    "\n",
    "print(all_plate_results_df.shape)\n",
    "all_plate_results_df.head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "6727b04d2b6bae3ded4fc387a1bb9b8f5bef8644bb0eed6fe44abf57f9b35522"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
